### ISSUE TITLE
gsva() causes crash on R v4.1.0

### ISSUE THREAD

**Issue by travis-m-blimkie** [rcastelo/GSVA#54]
Hello, 

I've recently been using this package as part of my RNA-Seq analysis, and encountered a problem since updating to R v4.1.0. When I try to run the main `gsva` function, R simply crashes/aborts during the process, without any error messages being printed - just the popup window stating that "R encountered a fatal error." But it runs fine with the exact same inputs if I switch back to R v4.0.5.  I'm using VST-normalized RNA-Seq data, with the following settings:

- method = "gsva"
- kcdf = "Gaussian"
- abs.ranking = TRUE

 Below is my `sessionInfo()` output. I can also provide some of my data/gene sets if that would be of use. 
```r
R version 4.1.0 (2021-05-18)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=English_Canada.1252  LC_CTYPE=English_Canada.1252   
[3] LC_MONETARY=English_Canada.1252 LC_NUMERIC=C                   
[5] LC_TIME=English_Canada.1252    

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] forcats_0.5.1        stringr_1.4.0        dplyr_1.0.6          purrr_0.3.4         
 [5] readr_1.4.0          tidyr_1.1.3          tibble_3.1.2         tidyverse_1.3.1     
 [9] plotly_4.9.3         ggplot2_3.3.3        DT_0.18              janitor_2.1.0       
[13] magrittr_2.0.1       ComplexHeatmap_2.8.0 circlize_0.4.12      GSVA_1.40.0         
[17] enrichR_3.0          ReactomePA_1.36.0    shinyjs_2.0.0        shiny_1.6.0
```
I was also able to reproduce the same behaviour on Ubuntu 20. Any help would be greatly appreciated. Thanks!

**Comment by rcastelo** [comment]
Hi, the package builds fine and passes all the preset tests, so I'd need to look at the specific data that triggers the error and have the specific instructions that you give, i.e., not just give the parameters you set, I need the actual line of code that executes the GSVA algorithm. You can contact me directly by email at robert.castelo@upf.edu

**Comment by travis-m-blimkie** [comment]
Thanks for the offer! There's a couple more things I'd like to try first, but if those don't resolve the issue I will send you the data/code over email. 

Cheers!

**Comment by travis-m-blimkie** [comment]
Alright, I seemed to have resolved the issue by removing genes with 0 variance across all samples prior to running `gsva()`. Which is a bit odd because this wasn't necessary when I was using the previous R version (4.0.5). But it's a pretty simple fix to implement. 

One final note that I'm not sure is helpful (or just adds to potential confusion lol...) -  it runs fine (without manually removing 0 variance genes on R v4.1.0) if I change the method to "ssgsea"... Just wanted to mention it in case it's helpful or if anyone else ends up here. 

Thanks again!

**Comment by rcastelo** [comment]
Hi,

That's correct, the `gsva()` function will remove genes with constant expression (i.e., rows for which the standard deviation is 0) as a first step, except when `method="ssgsea"`. The reason is that rows with a constant value are problematic in the computations made by all methods except in the case of ssGSEA, although in this case still I think you should carefully consider whether a gene with constant expression throughout samples provides more noise than signal to the gene sets to which it belongs to.

Could you tell me what was the exact line of code that in your computer generates the error and, at least, what class of objects and what dimensions do they have, in the input you are giving to the `gsva()` function? (I just run the last version of GSVA on a matrix of expression data with on row set to constant values and got no error, running on macOS)

**Comment by travis-m-blimkie** [comment]
Sure, the code is as follows:

```r
gsva(
    as.matrix(expr),
    gene_sets,
    method = "gsva",
    kcdf = "Gaussian",
    abs.ranking = TRUE
  )
```
`expr` is a dataframe with Ensembl gene IDs as rownames; there are 8 samples (columns) and 14,119 rows (14,112 after removing 0 variance genes). `gene_sets` is a named list of length 62, containing only gene sets with length >= 2.

I also triple checked that all genes IDs in the expression data and gene sets are not NA and match the regex "^ENSG". And I checked for any NA's in the data itself.

The reason I didn't include the data here initially is that it relates to some unpublished work, but I've got the OK to send it once I de-identify everything. I will email the expression data and gene sets, along with some code to test it out.

Cheers!

**Comment by rcastelo** [comment]
Hi,

found the problem, it's in fact caused by not effectively filtering all genes with constant expression. This is due to a change that we introduced in the last version of GSVA by which we would use the function `DelayedMatrixStats::rowSds()` instead of base `apply(x, 1, sd)` to be able to calculate the standard deviation on the rows of different classes of input matrices with a single function call. Unfortunately, it seems that `DelayedMatrixStats::rowSds()` calls `matrixStats::Sds()` and the latter seems to give values very close to zero, such as `2.340556e-16` where base `sd()` gives an exact zero and that allowed some genes with constant expression to sneak into the GSVA algorithm whose C code crashes with such genes. I have posted an [issue](https://github.com/HenrikBengtsson/matrixStats/issues/204) on the matrixStats package to see whether this has some solution on that end, but in the meantime I've submitted a temporary bugfix for the development and release version of GSVA (1.40.1), which should become available in the next 24/48 hours and which runs smoothly your code.

Thanks for catching this bug and for your help providing the necessary example data to figure it out!!

robert.
